source/common/http/http2/codec_impl.cc

```diff
@@ -483,12 +483,18 @@ int ConnectionImpl::onFrameSend(const nghttp2_frame* frame) {
   return 0;
 }
 
-int ConnectionImpl::onInvalidFrame(int error_code) {
+int ConnectionImpl::onInvalidFrame(int32_t stream_id, int error_code) {
   ENVOY_CONN_LOG(debug, "invalid frame: {}", connection_, nghttp2_strerror(error_code));
 
-  // The stream is about to be closed due to an invalid header. Don't kill the
-  // entire connection if one stream has bad headers.
-  if (error_code == NGHTTP2_ERR_HTTP_HEADER) {
+  // The stream is about to be closed due to an invalid header or messaging. Don't kill the
+  // entire connection if one stream has bad headers or messaging.
+  if (error_code == NGHTTP2_ERR_HTTP_HEADER || error_code == NGHTTP2_ERR_HTTP_MESSAGING) {
```

Comment:
Does nghttp2 save us from incomplete hpack state?  If the stream is closed based on an error in an initial headers frame but there's a contination, does it do the right thing or can we get wedged?  Worth adding a test for multi-frame request headers?

Reply:
I'm fairly certain it will do the right thing. Let me see about adding a test case for that. I'm not sure how hard it will be as nghttp2 generally hides continuation frames from us.

Reply:
HTTP semantics error occurs after successful HPACK decoding and its state is completely fine even with those errors.

Reply:
Great to hear, thank you for chiming in.

Curious, does that mean if the request headers are broken into multiple frames, we'll not send the stream cancellation until processing all of the headers, or we kill the stream off right away but continue to update HPACK state for internal consistency?

Reply:
Yes.  Unless we get unrecoverable error during HPACK decoding, we will feed incoming bits into HPACK even if its output is just thrown away.